File: zig/src/spec.zig
const std = @import("std");

pub const ParseError = error{
    Empty,
    BadCrateVersion,
    BadCrateName,
    BadPathSegment,
    OutOfMemory,
};

/// Parsed `crate[@version][::path::to::item]` reference.
/// Field slices borrow from the `raw` input passed to `parse`; keep it alive.
pub const ItemSpec = struct {
    crate_name: []const u8,
    version: []const u8,
    target: ?[]const u8 = null,
    path: [][]const u8,
    allocator: std.mem.Allocator,

    pub fn deinit(self: *ItemSpec) void {
        self.allocator.free(self.path);
    }

    pub fn isRoot(self: *const ItemSpec) bool {
        return self.path.len == 0;
    }

    pub fn parse(allocator: std.mem.Allocator, raw: []const u8) ParseError!ItemSpec {
        const trimmed = std.mem.trim(u8, raw, &std.ascii.whitespace);
        if (trimmed.len == 0) return ParseError.Empty;

        var head: []const u8 = trimmed;
        var rest: []const u8 = "";
        if (std.mem.indexOf(u8, trimmed, "::")) |i| {
            head = trimmed[0..i];
            rest = trimmed[i + 2 ..];
        }

        var path_list: std.ArrayList([]const u8) = .empty;
        errdefer path_list.deinit(allocator);

        if (rest.len > 0) {
            var it = std.mem.splitSequence(u8, rest, "::");
            while (it.next()) |seg| {
                if (!isValidIdent(seg)) return ParseError.BadPathSegment;
                try path_list.append(allocator, seg);
            }
        }

        var crate_name: []const u8 = head;
        var version: []const u8 = "latest";
        if (std.mem.indexOfScalar(u8, head, '@')) |i| {
            const c = head[0..i];
            const v = head[i + 1 ..];
            if (c.len == 0 or v.len == 0) return ParseError.BadCrateVersion;
            crate_name = c;
            version = v;
        }

        if (!isValidCrateName(crate_name)) return ParseError.BadCrateName;

        return ItemSpec{
            .crate_name = crate_name,
            .version = version,
            .path = try path_list.toOwnedSlice(allocator),
            .allocator = allocator,
        };
    }
};

const MAX_CRATE_NAME_LEN: usize = 64;

fn isValidCrateName(s: []const u8) bool {
    if (s.len == 0 or s.len > MAX_CRATE_NAME_LEN) return false;
    if (!std.ascii.isAlphabetic(s[0])) return false;
    for (s[1..]) |c| {
        if (!(std.ascii.isAlphanumeric(c) or c == '-' or c == '_')) return false;
    }
    return true;
}

fn isValidIdent(s: []const u8) bool {
    if (s.len == 0) return false;
    for (s) |c| {
        if (!(std.ascii.isAlphanumeric(c) or c == '_')) return false;
    }
    return true;
}

test "bare crate" {
    const a = std.testing.allocator;
    var s = try ItemSpec.parse(a, "serde");
    defer s.deinit();
    try std.testing.expectEqualStrings("serde", s.crate_name);
    try std.testing.expectEqualStrings("latest", s.version);
    try std.testing.expect(s.isRoot());
}

test "crate with version" {
    const a = std.testing.allocator;
    var s = try ItemSpec.parse(a, "serde@1.0.200");
    defer s.deinit();
    try std.testing.expectEqualStrings("1.0.200", s.version);
}

test "crate with path" {
    const a = std.testing.allocator;
    var s = try ItemSpec.parse(a, "serde::de::Deserialize");
    defer s.deinit();
    try std.testing.expectEqualStrings("serde", s.crate_name);
    try std.testing.expectEqual(@as(usize, 2), s.path.len);
    try std.testing.expectEqualStrings("de", s.path[0]);
    try std.testing.expectEqualStrings("Deserialize", s.path[1]);
}

test "crate with version and path" {
    const a = std.testing.allocator;
    var s = try ItemSpec.parse(a, "anyhow@~1::Error");
    defer s.deinit();
    try std.testing.expectEqualStrings("anyhow", s.crate_name);
    try std.testing.expectEqualStrings("~1", s.version);
    try std.testing.expectEqual(@as(usize, 1), s.path.len);
    try std.testing.expectEqualStrings("Error", s.path[0]);
}

test "rejects empty" {
    try std.testing.expectError(ParseError.Empty, ItemSpec.parse(std.testing.allocator, ""));
}

test "rejects bad at" {
    try std.testing.expectError(ParseError.BadCrateVersion, ItemSpec.parse(std.testing.allocator, "@1.0"));
    try std.testing.expectError(ParseError.BadCrateVersion, ItemSpec.parse(std.testing.allocator, "serde@"));
}

test "rejects invalid crate names" {
    const a = std.testing.allocator;
    try std.testing.expectError(ParseError.BadCrateName, ItemSpec.parse(a, "1serde"));
    try std.testing.expectError(ParseError.BadCrateName, ItemSpec.parse(a, "-serde"));
    try std.testing.expectError(ParseError.BadCrateName, ItemSpec.parse(a, "_serde"));
    try std.testing.expectError(ParseError.BadCrateName, ItemSpec.parse(a, "ser de"));
    try std.testing.expectError(ParseError.BadCrateName, ItemSpec.parse(a, "ser.de"));
    const long = "a" ** 65;
    try std.testing.expectError(ParseError.BadCrateName, ItemSpec.parse(a, long));
}

test "accepts valid crate names" {
    const a = std.testing.allocator;
    const max_name = "a" ** 64;
    inline for (.{ "serde", "serde_json", "tracing-subscriber", "a", max_name }) |name| {
        var s = try ItemSpec.parse(a, name);
        defer s.deinit();
    }
}

File: src/spec.rs
use crate::{Error, Result};

/// A parsed reference to a rustdoc item on docs.rs.
///
/// Examples of accepted grammar:
///   - `serde`
///   - `serde@1.0.200`
///   - `serde::de::Deserialize`
///   - `serde@~1::Serializer`
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ItemSpec {
    pub crate_name: String,
    /// Version string as accepted by docs.rs: "latest", "1.0.200", "~1", etc.
    pub version: String,
    /// Rustdoc target triple; `None` means the default host target.
    pub target: Option<String>,
    /// Path components *without* the crate name, e.g. `["de", "Deserialize"]`.
    /// Empty means the crate root module.
    pub path: Vec<String>,
}

impl ItemSpec {
    /// # Errors
    /// Returns `Error::InvalidSpec` when `raw` is empty, when the
    /// `crate@version` portion is malformed, or when any segment fails the
    /// crate-name / identifier validity rules.
    pub fn parse(raw: &str) -> Result<Self> {
        let raw = raw.trim();
        if raw.is_empty() {
            return Err(Error::InvalidSpec("empty".into()));
        }

        // Split off the path on `::`. The first segment is `crate[@version]`.
        let (head, rest) = raw.split_once("::").unwrap_or((raw, ""));
        let path: Vec<String> = if rest.is_empty() {
            Vec::new()
        } else {
            rest.split("::").map(str::to_string).collect()
        };

        let (crate_name, version) = match head.split_once('@') {
            Some((c, v)) if !c.is_empty() && !v.is_empty() => (c.to_string(), v.to_string()),
            Some(_) => return Err(Error::InvalidSpec(format!("bad crate@version: {head}"))),
            None => (head.to_string(), "latest".to_string()),
        };

        if !is_valid_crate_name(&crate_name) {
            return Err(Error::InvalidSpec(format!("bad crate name: {crate_name}")));
        }
        for seg in &path {
            if !is_valid_ident(seg) {
                return Err(Error::InvalidSpec(format!("bad path segment: {seg}")));
            }
        }

        Ok(Self {
            crate_name,
            version,
            target: None,
            path,
        })
    }

    #[must_use]
    pub fn with_target(mut self, target: Option<String>) -> Self {
        self.target = target;
        self
    }

    #[must_use]
    pub fn is_root(&self) -> bool {
        self.path.is_empty()
    }

    /// Full path as rustdoc sees it: `[crate_name, ..path]`.
    #[must_use]
    pub fn full_path(&self) -> Vec<String> {
        let mut out = Vec::with_capacity(self.path.len() + 1);
        out.push(self.crate_name.clone());
        out.extend(self.path.iter().cloned());
        out
    }
}

/// Matches the crates.io/docs.rs rules:
/// non-empty, <= 64 chars, first char ASCII alphabetic, rest ASCII alphanumeric or `-`/`_`.
/// See <https://github.com/rust-lang/docs.rs/blob/main/crates/lib/crates_io_validation/src/lib.rs>
fn is_valid_crate_name(s: &str) -> bool {
    const MAX_NAME_LENGTH: usize = 64;
    if s.is_empty() || s.chars().count() > MAX_NAME_LENGTH {
        return false;
    }
    let mut chars = s.chars();
    let Some(first) = chars.next() else {
        return false;
    };
    if !first.is_ascii_alphabetic() {
        return false;
    }
    chars.all(|c| c.is_ascii_alphanumeric() || c == '-' || c == '_')
}

fn is_valid_ident(s: &str) -> bool {
    !s.is_empty() && s.chars().all(|c| c.is_ascii_alphanumeric() || c == '_')
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn bare_crate() {
        let s = ItemSpec::parse("serde").unwrap();
        assert_eq!(s.crate_name, "serde");
        assert_eq!(s.version, "latest");
        assert!(s.path.is_empty());
    }

    #[test]
    fn crate_with_version() {
        let s = ItemSpec::parse("serde@1.0.200").unwrap();
        assert_eq!(s.version, "1.0.200");
    }

    #[test]
    fn crate_with_path() {
        let s = ItemSpec::parse("serde::de::Deserialize").unwrap();
        assert_eq!(s.crate_name, "serde");
        assert_eq!(s.version, "latest");
        assert_eq!(s.path, vec!["de", "Deserialize"]);
    }

    #[test]
    fn crate_with_version_and_path() {
        let s = ItemSpec::parse("anyhow@~1::Error").unwrap();
        assert_eq!(s.crate_name, "anyhow");
        assert_eq!(s.version, "~1");
        assert_eq!(s.path, vec!["Error"]);
    }

    #[test]
    fn rejects_empty() {
        assert!(ItemSpec::parse("").is_err());
    }

    #[test]
    fn rejects_bad_at() {
        assert!(ItemSpec::parse("@1.0").is_err());
        assert!(ItemSpec::parse("serde@").is_err());
    }

    #[test]
    fn rejects_invalid_crate_names() {
        assert!(ItemSpec::parse("1serde").is_err(), "leading digit");
        assert!(ItemSpec::parse("-serde").is_err(), "leading dash");
        assert!(ItemSpec::parse("_serde").is_err(), "leading underscore");
        assert!(ItemSpec::parse("ser de").is_err(), "space");
        assert!(ItemSpec::parse("ser.de").is_err(), "dot");
        assert!(ItemSpec::parse(&"a".repeat(65)).is_err(), "too long");
    }

    #[test]
    fn accepts_valid_crate_names() {
        assert!(ItemSpec::parse("serde").is_ok());
        assert!(ItemSpec::parse("serde_json").is_ok());
        assert!(ItemSpec::parse("tracing-subscriber").is_ok());
        assert!(ItemSpec::parse("a").is_ok());
        assert!(ItemSpec::parse(&"a".repeat(64)).is_ok(), "max length");
    }
}
